src/dbnode/storage/bootstrap_instrumentation.go

```diff
@@ -0,0 +1,167 @@
+	i.logFields = append(i.logFields, zap.Int("numShards", shards))
+	i.log.Info("bootstrap started", i.logFields...)
+}
+
+func (i *instrumentationContext) bootstrapSucceeded() {
+	bootstrapDuration := i.nowFn().Sub(i.start)
+	i.bootstrapDuration.Record(bootstrapDuration)
+	i.logFields = append(i.logFields, zap.Duration("bootstrapDuration", bootstrapDuration))
+	i.log.Info("bootstrap succeeded, marking namespaces complete", i.logFields...)
+}
+
+func (i *instrumentationContext) bootstrapFailed(err error) {
+	bootstrapDuration := i.nowFn().Sub(i.start)
+	i.bootstrapDuration.Record(bootstrapDuration)
+	i.logFields = append(i.logFields, zap.Duration("bootstrapDuration", bootstrapDuration))
+	i.log.Error("bootstrap failed", append(i.logFields, zap.Error(err))...)
+}
+
+func (i *instrumentationContext) bootstrapNamespacesStarted() {
+	i.start = i.nowFn()
+	i.log.Info("bootstrap namespaces start", i.logFields...)
+}
+
+func (i *instrumentationContext) bootstrapNamespacesSucceeded() {
+	duration := i.nowFn().Sub(i.start)
+	i.bootstrapNamespacesDuration.Record(duration)
+	i.logFields = append(i.logFields, zap.Duration("bootstrapNamespacesDuration", duration))
+	i.log.Info("bootstrap namespaces success", i.logFields...)
+}
+
+func (i *instrumentationContext) bootstrapNamespaceFailed(
+	err error,
+	namespaceID ident.ID,
+) {
+	i.log.Info("bootstrap namespace error", append(i.logFields,
+		zap.String("namespace", namespaceID.String()),
+		zap.Error(err))...)
+}
+
+func (i *instrumentationContext) bootstrapNamespacesFailed(err error) {
+	duration := i.nowFn().Sub(i.start)
+	i.bootstrapNamespacesDuration.Record(duration)
+	i.logFields = append(i.logFields, zap.Duration("bootstrapNamespacesDuration", duration))
+	i.log.Info("bootstrap namespaces failed", append(i.logFields, zap.Error(err))...)
+}
+
+func (i *instrumentationContext) logFn(err error, msg string) func(l *zap.Logger) {
+	return func(l *zap.Logger) {
+		l.Error(msg, append(i.logFields, zap.Error(err))...)
+	}
+}
+
+type bootstrapInstrumentation struct {
+	opts          Options
+	scope         tally.Scope
+	log           *zap.Logger
+	nowFn         clock.NowFn
+	status        tally.Gauge
+	durableStatus tally.Gauge
+	numRetries    tally.Counter
+}
+
+func newBootstrapInstrumentation(opts Options) *bootstrapInstrumentation {
+	scope := opts.InstrumentOptions().MetricsScope()
+	return &bootstrapInstrumentation{
+		opts:          opts,
+		scope:         scope,
+		log:           opts.InstrumentOptions().Logger(),
+		nowFn:         opts.ClockOptions().NowFn(),
+		status:        scope.Gauge("bootstrapped"),
+		durableStatus: scope.Gauge("bootstrapped-durable"),
+		numRetries:    scope.Counter("bootstrap-retries"),
+	}
+}
+
+func (i *bootstrapInstrumentation) bootstrapPreparing() *instrumentationContext {
+	i.log.Info("bootstrap prepare")
+	return newInstrumentationContext(i.nowFn, i.log, i.scope, i.opts)
+}
+
+func (i *bootstrapInstrumentation) bootstrapFnFailed(retry int) {
```

Comment:
Just call this `bootstrapFailed` perhaps to be consistent with the other naming? Doesn't look like the `fn` is passed anywhere as a parameter or anything, so it shouldn't be any different to the other methods on this struct?